ichnaea/locate/tests/test_provider.py

```diff
@@ -700,3 +705,32 @@ def test_should_not_provide_location_if_non_geoip_location_found(self):
         }
         self.assertFalse(
             self.provider.should_locate(query_data, internal_location))
+
+    def test_rate_limiting_restricts_calls(self):
+        response_location = {
+            'location': {
+                'lat': 51.5366,
+                'lng': 0.03989,
+            },
+            'accuracy': 1500,
+            'fallback': 'lacf',
+        }
+
+        with requests_mock.Mocker() as mock:
+            mock.register_uri(
+                'POST', requests_mock.ANY, json=response_location)
+
+            for _ in range(self.provider.settings['ratelimit']):
+                location = self.provider.locate({
+                    'cell': self.cells,
+                    'wifi': self.wifis,
+                })
+
+                self.assertTrue(location.found())
+
+            location = self.provider.locate({
+                'cell': self.cells,
+                'wifi': self.wifis,
+            })
+
+            self.assertFalse(location.found())
```

Comment:
This assumes this call is happening in the same second, as the successful attempts earlier. If we are unlucky, the second will roll over right in between these calls.

I'd rather make this a bit more stable:
* Have one test that checks it can do queries up until incl. the rate limit.
* Add a second test that manually constructs and adds the rate limit exceeded keys for the next three seconds into Redis, than calls the provider and checks that it respected the limit. By adding the keys for a couple continuous seconds, we should be safe from second roll-over in the middle of it.